### PR TITLE
[Python] Convert DECIMAL values to `decimal.Decimal` in NumPy/Pandas output

### PR DESCRIPTION
**Pull request by Tishj**
Previously we converted DECIMAL values into `double` when outputting to NumPy (and by association - Pandas)
This is not lossless, so this causes issues in the long run.

When you use a DECIMAL type you usually care about the precision it provides, so losing this is not ideal.
If you want to preserve the old behavior, you can push a cost to DOUBLE manually.

### `test_all_types` test fixes

While working on this, I noticed that the numpy `test_all_types.py` test portion was broken.
I fixed up the test and had to comment out some tests for arrays.

The reason these have to be commented out is that we use `numpy.testing.assert_equal`, which throws an error when empty arrays are tested.
Our `test_all_types` output for arrays contains empty arrays so this is problematic.

### CODE REVIEW

File: scripts/regression_test_python.py [hunk 1]
```diff
@@ -72,6 +72,8 @@ def __init__(self, scale_factor):
 
     def get_tables(self, convertor) -> Dict[str, Any]:
         res = {}
+        # Without this, the DECIMAL type analyzed could be too small depending on the distribution
+        self.conn.execute("set global pandas_analyze_sample=2048")
```
Comment (Tishj) [inline scripts/regression_test_python.py:76]: This is caused by the pandas analyzer not seeing decimals with a big enough width.
One other way we can fix this is by making the detected type by the analyzer wider

So whenever we find a DECIMAL with a width that would use a `int16_t` internally, we would set the detected width to the max of that physical type (4 in this case)

File: tools/pythonpkg/src/numpy/array_wrapper.cpp [hunk 2]
```diff
@@ -572,26 +572,44 @@ static bool ConvertColumnRegular(NumpyAppendData &append_data) {
...
 
 template <class DUCKDB_T>
-static bool ConvertDecimalInternal(NumpyAppendData &append_data, double division) {
+static PyObject *ConstructDecimal(DUCKDB_T value, uint8_t width, uint8_t scale) {
+	auto decimal_length = DecimalToString::DecimalLength<DUCKDB_T>(value, width, scale);
+
+	auto &python_import_cache = *DuckDBPyConnection::ImportCache();
+	auto decimal = python_import_cache.decimal.Decimal();
+
+	unique_ptr<char[]> raw_string(new char[decimal_length]);
```
Comment (Tishj) [inline tools/pythonpkg/src/numpy/array_wrapper.cpp:581]: This looks medium hacky,`decimal_length` is variable based on the `value` so we can't reuse the buffers
We might be able to find the max length given the physical type and reuse the buffer that way.

File: scripts/regression_test_python.py (hunk 1 above)
Comment (Tishj) [reply scripts/regression_test_python.py:76]: Seems that this bandaid doesn't work even, I'll try the widening, that sounds like a better solution in general

**Comment by Tishj** [comment]
```
pandastpch
Old timing: 5.567067146301[26](https://github.com/Tishj/duckdb/actions/runs/8910519790/job/24469965760#step:11:27)95
New timing: 808.2724604606628
```

I outdid myself, a 16x regression

File: tools/pythonpkg/src/numpy/array_wrapper.cpp (hunk 2 above)
Comment (Tishj) [reply tools/pythonpkg/src/numpy/array_wrapper.cpp]: After seeing the enormous regression I figured I would reuse the string buffer and see how much that helps
From 808 -> 802 seconds, so.. not much

I'll have to profile it but I think the problem is just the constructor of the decimal.Decimal objects, to which there isn't much that can be done.

We could instead keep the old conversion and put this one behind a flag??
`python_preserve_decimal` ?

File: tools/pythonpkg/src/numpy/array_wrapper.cpp [hunk 3]
```diff
@@ -572,26 +572,45 @@ static bool ConvertColumnRegular(NumpyAppendData &append_data) {
...
-static bool ConvertDecimalInternal(NumpyAppendData &append_data, double division) {
+static PyObject *ConstructDecimal(DUCKDB_T value, uint8_t width, uint8_t scale, char *buffer) {
+	auto decimal_length = DecimalToString::DecimalLength<DUCKDB_T>(value, width, scale);
+
+	auto &python_import_cache = *DuckDBPyConnection::ImportCache();
+	auto decimal = python_import_cache.decimal.Decimal();
+
+	DecimalToString::FormatDecimal<DUCKDB_T>(value, width, scale, buffer, UnsafeNumericCast<idx_t>(decimal_length));
+	auto stringified = string(buffer, decimal_length);
+	auto py_obj = decimal(py::str(stringified));
```
Comment (Tishj) [inline tools/pythonpkg/src/numpy/array_wrapper.cpp:583]: Another way to construct a `decimal.Decimal` is by using the `tuple` constructor:

Example to create a decimal `1.234`:
`decimal.Decimal((0, (1, 2, 3, 4), -3))`

That would require duplicating a lot of the decimal conversion logic to create the digits tuple `(1, 2, 3, 4)` and involves creating even more python objects.
I haven't tested this but I can't imagine this being faster than the current approach

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Let's wait on this given the large performance regression. Maybe if someone requests it we can put this behind a feature flag.

**Comment by Tishj** [comment]
Sounds good, I'll close it for now then

**Pull request closed without merging**